tests/phpunit/includes/bootstrap.php

```diff
@@ -46,13 +46,83 @@
-if ( ! file_exists( $phpunit_polyfills_autoloader ) ) {
-	echo 'Error: You need to run `composer update` before running the tests.' . PHP_EOL;
-	echo 'You can still use a PHPUnit phar to run them, but the dependencies do need to be installed.' . PHP_EOL;
+/*
+ * Load the PHPUnit Polyfills autoloader.
+ *
+ * The PHPUnit Polyfills are a requirement for the WP Test Suite.
+ *
+ * Plugin/theme integration tests can handle this in any of the following ways:
+ * - When using a full WP install: run `composer update` for the WP install prior to running the tests.
+ * - When using a partial WP test suite install:
+ *   - Add a `yoast/phpunit-polyfills:"^1.0.1"` requirement to their own `composer.json` file.
+ *   - And then:
+ *     - Either load the PHPUnit Polyfills autoload file prior to running the WP Core bootstrap file.
+ *     - Or declare a `WP_TESTS_PHPUNIT_POLYFILLS_PATH` constant pointing to the root directory
+ *       of the PHPUnit Polyfills installation.
+ *       This constant can be declared in the `phpunit.xml[.dist]` file like so:
+ *       `<php><const name="WP_TESTS_PHPUNIT_POLYFILLS_PATH" value="path/to/yoast/phpunit-polyfills"/></php>
+ *       or can be declared as a PHP constant in the `wp-tests-config.php` file or the plugin/theme
+ *       test bootstrap file.
+ */
+if ( ! class_exists( 'Yoast\PHPUnitPolyfills\Autoload' ) ) {
+	// Default location of the autoloader for WP Core test runs.
+	$phpunit_polyfills_autoloader = __DIR__ . '/../../../vendor/yoast/phpunit-polyfills/phpunitpolyfills-autoload.php';
+	$phpunit_polyfills_error      = false;
+
+	// Allow for a custom installation location to be provided for plugin/theme integration tests.
+	if ( defined( 'WP_TESTS_PHPUNIT_POLYFILLS_PATH' ) ) {
+		$phpunit_polyfills_path = WP_TESTS_PHPUNIT_POLYFILLS_PATH;
+
+		if ( is_string( WP_TESTS_PHPUNIT_POLYFILLS_PATH )
+			&& '' !== WP_TESTS_PHPUNIT_POLYFILLS_PATH
+		) {
+			$phpunit_polyfills_path = rtrim( $phpunit_polyfills_path, '/\\' );
+			$phpunit_polyfills_path = realpath( $phpunit_polyfills_path . '/phpunitpolyfills-autoload.php' );
+			if ( false !== $phpunit_polyfills_path ) {
+				$phpunit_polyfills_autoloader = $phpunit_polyfills_path;
+			} else {
+				$phpunit_polyfills_error = true;
+			}
+		} else {
+			$phpunit_polyfills_error = true;
+		}
+	}
+
+	if ( $phpunit_polyfills_error || ! file_exists( $phpunit_polyfills_autoloader ) ) {
+		echo 'Error: The PHPUnit Polyfills library is a requirement for running the WP test suite.' . PHP_EOL;
+		if ( isset( $phpunit_polyfills_path ) ) {
+			printf(
+				'The PHPUnit Polyfills autoload file was not found in "%s"' . PHP_EOL,
+				WP_TESTS_PHPUNIT_POLYFILLS_PATH
+			);
+			echo 'Please verify that the file path provided in the WP_TESTS_PHPUNIT_POLYFILLS_PATH constant is correct.' . PHP_EOL;
+		} else {
+			echo 'You need to run `composer update` before running the tests.' . PHP_EOL;
```

Comment:
When running the test suite without first doing `composer install`, there are 2 different experiences:

**Scenario 1:**

- No `vendor` folder as `composer install` not ran
- `WP_TESTS_PHPUNIT_POLYFILLS_PATH` constant not defined

Results:
```
Error: The PHPUnit Polyfills library is a requirement for running the WP test suite.
You need to run `composer update` before running the tests.
You can still use a PHPUnit phar to run them, but the dependencies do need to be installed.
```

**Scenario 2:**

- No `vendor` folder as `composer install` not ran
- `WP_TESTS_PHPUNIT_POLYFILLS_PATH` constant defined to a valid directory that does not yet have the `vendor` folder in it as `composer install` has not yet been run

```php
define( 'WP_TESTS_PHPUNIT_POLYFILLS_PATH', './valid_dir/vendor/yoast/phpunit-polyfills/' );
```

Results:
```
Error: The PHPUnit Polyfills library is a requirement for running the WP test suite.
The PHPUnit Polyfills autoload file was not found in "./valid_dir/vendor/yoast/phpunit-polyfills/"
Please verify that the file path provided in the WP_TESTS_PHPUNIT_POLYFILLS_PATH constant is correct.
```

Notice how Scenario 2 does not guide the extender to run `composer install` or `composer update`. 

@jrfnl What do you think? Should it be more descriptive?

@jrfnl What do you think? Should it be more descriptive?

Reply:
Updated: Moved the comment here for inline discussion relative to the code.

Reply:
I did it like this **_on purpose_** as, in the case the constant is used, WordPress has no awareness of the type of installation (and shouldn't care either), so we cannot presume that it will be a Composer install.

The constant is expected to be pointed to the _installation directory_ for the PHPUnit Polyfills.
This can be - and most likely will be - a Composer `vendor/...` directory, but it can also be a stand-alone install of the PHPUnit Polyfills (tested and works just fine), a directory containing a copied in version of the Polyfills (not recommended, violates copyright, but will work), a custom install directory or even a Composer global install directory.

It is not for WordPress to dictate how a user who chooses to use the `WP_TESTS_PHPUNIT_POLYFILLS_PATH` constant organizes their development environment, so no assumptions about this should be made in the message either.


Also notice that there is still a third scenario for which the same message from scenario 2 will be shown:

#### Scenario 3:
*  `vendor` folder exist, `composer install` or `update` has been run.
* `WP_TESTS_PHPUNIT_POLYFILLS_PATH` constant defined, but contains a typo or otherwise points to the wrong directory, like when it would be pointed to the `vendor` directory without the subpath.

Reply:
Cool that makes sense!

Reply:
I have a feeling that inline code comment are not posted through to the Trac ticket. Shall I copy this comment to the "main" thread so it will be visible in the Trac ticket ? That would document the conversation in Trac better.